File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.java
package net.squanchy.navigation.firststart;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.graphics.drawable.Drawable;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.FastOutLinearInInterpolator;
import android.util.Property;
import android.view.View;
import android.view.animation.BounceInterpolator;
import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;
import android.widget.ImageView;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.fonts.TypefaceStyleableActivity;
import net.squanchy.support.config.DialogLayoutParameters;
import net.squanchy.support.lang.Func1;

import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import timber.log.Timber;

@SuppressWarnings("PMD.ExcessiveImports")           // Might use some refactoring later on to extract some collaborator
public class FirstStartWithNoNetworkActivity extends TypefaceStyleableActivity {

    private static final String EXTRA_CONTINUATION_INTENT = FirstStartWithNoNetworkActivity.class.getCanonicalName() + ".continuation_intent";
    private static final long DELAY_AFTER_ANIMATIONS_MILLIS = 700;

    private Intent continuationIntent;

    private ConnectivityManager.NetworkCallback networkCallback;
    private ConnectivityManager connectivityManager;
    private Handler mainThreadHandler;

    private ImageView ctaProgressView;
    private TextView ctaTextView;
    private int shortAnimationDuration;

    public static Intent createIntentContinuingTo(Context context, Intent continuationIntent) {
        Intent intent = new Intent(context, FirstStartWithNoNetworkActivity.class);
        intent.putExtra(EXTRA_CONTINUATION_INTENT, continuationIntent);
        return intent;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_first_start_with_no_network);

        DialogLayoutParameters.wrapHeight(this)
                .applyTo(getWindow());

        findViewById(R.id.first_start_nevermind_button).setOnClickListener(view -> finish());
        ctaProgressView = (ImageView) findViewById(R.id.cta_progressbar);
        ctaTextView = (TextView) findViewById(R.id.cta_text);

        continuationIntent = getIntent().getParcelableExtra(EXTRA_CONTINUATION_INTENT);

        connectivityManager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        networkCallback = new NetworkConnectedCallback();
        mainThreadHandler = new Handler(Looper.getMainLooper());

        shortAnimationDuration = getResources().getInteger(android.R.integer.config_shortAnimTime);
    }

    @Override
    protected void onStart() {
        super.onStart();

        NetworkRequest networkRequest = new NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .build();
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback);

        doMaybeOnAnimatedVector(ctaProgressView.getDrawable(), AnimatedVectorDrawable::start);
    }

    @Override
    protected void onStop() {
        super.onStop();
        connectivityManager.unregisterNetworkCallback(networkCallback);
        doMaybeOnAnimatedVector(ctaProgressView.getDrawable(), AnimatedVectorDrawable::stop);
    }

    @SuppressWarnings("PMD.AvoidCatchingGenericException") // The Consumer signature has a throws Exception, we have to catch it
    private void doMaybeOnAnimatedVector(Drawable drawable, Consumer<AnimatedVectorDrawable> action) {
        if (drawable instanceof AnimatedVectorDrawable) {
            try {
                action.accept((AnimatedVectorDrawable) drawable);
            } catch (Exception e) {
                Timber.e(e);
            }
        }
    }

    private void showNetworkAcquiredAndFinish() {
        ctaTextView.setText(R.string.first_start_with_no_network_network_connected);

        animate(ctaProgressView, popOut(), this::swapProgressWithSuccessAndContinue)
                .start();
    }

    private Func1<View, Animator> popOut() {
        return view -> {
            Animator alphaAnimator = createFadeAnimationFor(view, 1f, 0f);
            Animator scaleAnimator = createScaleAnimationFor(
                    view,
                    new FastOutLinearInInterpolator(),
                    1f, 0f);
            AnimatorSet animator = new AnimatorSet();
            animator.playTogether(alphaAnimator, scaleAnimator);
            animator.setDuration(shortAnimationDuration);
            return animator;
        };
    }

    private void swapProgressWithSuccessAndContinue() {
        ctaProgressView.setImageResource(R.drawable.ic_circle_tick);

        animate(ctaProgressView, popBackIn(), this::continueToScheduleAfterDelay)
                .start();
    }

    private Func1<View, Animator> popBackIn() {
        return view -> {
            Animator alphaAnimator = createFadeAnimationFor(view, 0f, 1f);
            Animator scaleAnimator = createScaleAnimationFor(
                    view,
                    new BounceInterpolator(),
                    0f, 1f);
            AnimatorSet animator = new AnimatorSet();
            animator.playTogether(alphaAnimator, scaleAnimator);
            animator.setDuration(shortAnimationDuration);
            return animator;
        };
    }

    private Animator createFadeAnimationFor(View view, float... values) {
        Property<View, Float> property = Property.of(View.class, Float.class, "alpha");
        ObjectAnimator animator = ObjectAnimator.ofFloat(view, property, values);
        animator.setInterpolator(new LinearInterpolator());
        return animator;
    }

    private Animator createScaleAnimationFor(View view, Interpolator interpolator, float... values) {
        Property<View, Float> scaleX = Property.of(View.class, Float.class, "scaleX");
        Property<View, Float> scaleY = Property.of(View.class, Float.class, "scaleY");
        AnimatorSet animator = new AnimatorSet();
        animator.playTogether(
                ObjectAnimator.ofFloat(view, scaleX, values),
                ObjectAnimator.ofFloat(view, scaleY, values)
        );
        animator.setInterpolator(interpolator);
        return animator;
    }

    private Animator animate(View view, Func1<View, Animator> animationProducer, Action endAction) {
        Animator animator = animationProducer.call(view);
        animator.addListener(new AnimationEndListener() {
            @Override
            @SuppressWarnings("PMD.AvoidCatchingGenericException") // The Action signature has a throws Exception, we have to catch it
            public void onAnimationEnd(Animator animation) {
                try {
                    endAction.run();
                } catch (Exception e) {
                    Timber.e(e);
                }
            }
        });
        return animator;
    }

    private void continueToScheduleAfterDelay() {
        ctaProgressView.postDelayed(() -> {
            startActivity(continuationIntent);      // We don't use the navigator here, we basically want to restart the whole flow
            finish();
        }, DELAY_AFTER_ANIMATIONS_MILLIS);
    }

    private class NetworkConnectedCallback extends ConnectivityManager.NetworkCallback {

        private boolean receivedOnAvailable;

        @Override
        public void onAvailable(Network network) {
            if (receivedOnAvailable) {
                return;
            }

            receivedOnAvailable = true;

            mainThreadHandler.post(FirstStartWithNoNetworkActivity.this::showNetworkAcquiredAndFinish);
        }
    }

    private static class AnimationEndListener implements Animator.AnimatorListener {

        @Override
        public final void onAnimationStart(Animator animation) {
            // No-op
        }

        @Override
        public void onAnimationEnd(Animator animation) {
            // To override
        }

        @Override
        public final void onAnimationCancel(Animator animation) {
            // No-op
        }

        @Override
        public final void onAnimationRepeat(Animator animation) {
            // No-op
        }
    }
}

File: app/src/main/java/net/squanchy/navigation/firststart/FirstStartWithNoNetworkActivity.kt
package net.squanchy.navigation.firststart

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.content.Intent
import android.graphics.drawable.AnimatedVectorDrawable
import android.graphics.drawable.Drawable
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Property
import android.view.View
import android.view.animation.BounceInterpolator
import android.view.animation.Interpolator
import android.view.animation.LinearInterpolator
import androidx.appcompat.app.AppCompatActivity
import androidx.interpolator.view.animation.FastOutLinearInInterpolator
import kotlinx.android.synthetic.main.activity_first_start_with_no_network.*
import net.squanchy.R
import net.squanchy.support.config.DialogLayoutParameters

class FirstStartWithNoNetworkActivity : AppCompatActivity() {

    private lateinit var continuationIntent: Intent

    private var networkCallback = NetworkConnectedCallback()
    private var mainThreadHandler = Handler(Looper.getMainLooper())

    private lateinit var connectivityManager: ConnectivityManager

    private var shortAnimationDuration: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_first_start_with_no_network)

        DialogLayoutParameters.wrapHeight(this)
            .applyTo(window)

        firstStartNevermind.setOnClickListener { finish() }

        continuationIntent = intent.getParcelableExtra(EXTRA_CONTINUATION_INTENT)

        connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

        shortAnimationDuration = resources.getInteger(android.R.integer.config_shortAnimTime)
    }

    override fun onStart() {
        super.onStart()

        val networkRequest = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        connectivityManager.registerNetworkCallback(networkRequest, networkCallback)

        doIfAnimatedVectorDrawable(firstStartProgress.drawable) { start() }
    }

    override fun onStop() {
        super.onStop()
        connectivityManager.unregisterNetworkCallback(networkCallback)
        doIfAnimatedVectorDrawable(firstStartProgress.drawable) { stop() }
    }

    private fun doIfAnimatedVectorDrawable(drawable: Drawable, action: AnimatedVectorDrawable.() -> Unit) {
        if (drawable is AnimatedVectorDrawable) {
            action.invoke(drawable)
        }
    }

    private fun showNetworkAcquiredAndFinish() {
        firstStartCta.setText(R.string.first_start_with_no_network_network_connected)

        animate(firstStartProgress, popOut(), ::swapProgressWithSuccessAndContinue)
            .start()
    }

    private fun popOut(): (View) -> Animator = {
        AnimatorSet().apply {
            val alphaAnimator = createFadeAnimationFor(it, 1f, 0f)
            val scaleAnimator = createScaleAnimationFor(it, FastOutLinearInInterpolator(), 1f, 0f)
            playTogether(alphaAnimator, scaleAnimator)
            duration = shortAnimationDuration.toLong()
        }
    }

    private fun swapProgressWithSuccessAndContinue() {
        firstStartProgress.setImageResource(R.drawable.ic_circle_tick)

        animate(firstStartProgress, popBackIn()) { this.continueToScheduleAfterDelay() }
            .start()
    }

    private fun popBackIn(): (View) -> Animator = {
        AnimatorSet().apply {
            val alphaAnimator = createFadeAnimationFor(it, 0f, 1f)
            val scaleAnimator = createScaleAnimationFor(it, BounceInterpolator(), 0f, 1f)
            playTogether(alphaAnimator, scaleAnimator)
            duration = shortAnimationDuration.toLong()
        }
    }

    @Suppress("SpreadOperator") // We cannot avoid using the spread operator here, we use varargs APIs
    private fun createFadeAnimationFor(view: View, vararg values: Float): Animator {
        val property = Property.of(View::class.java, Float::class.java, "alpha")
        val animator = ObjectAnimator.ofFloat(view, property, *values)
        animator.interpolator = LinearInterpolator()
        return animator
    }

    @Suppress("SpreadOperator") // We cannot avoid using the spread operator here, we use varargs APIs
    private fun createScaleAnimationFor(view: View, interpolator: Interpolator, vararg values: Float): Animator {
        val scaleX = Property.of(View::class.java, Float::class.java, "scaleX")
        val scaleY = Property.of(View::class.java, Float::class.java, "scaleY")
        val animator = AnimatorSet()
        animator.playTogether(
                ObjectAnimator.ofFloat(view, scaleX, *values),
                ObjectAnimator.ofFloat(view, scaleY, *values)
        )
        animator.interpolator = interpolator
        return animator
    }

    private fun animate(view: View, animationProducer: (View) -> Animator, endAction: () -> Unit): Animator {
        val animator = animationProducer.invoke(view)
        animator.addListener(
                object : AnimationEndListener {
                    override fun onAnimationEnd(animation: Animator) = endAction()
                }
        )
        return animator
    }

    private fun continueToScheduleAfterDelay() {
        firstStartProgress.postDelayed(
                {
                    startActivity(continuationIntent) // We don't use the navigator here, we basically want to restart the whole flow
                    finish()
                }, DELAY_AFTER_ANIMATIONS_MILLIS
        )
    }

    private inner class NetworkConnectedCallback : ConnectivityManager.NetworkCallback() {

        private var receivedOnAvailable: Boolean = false

        override fun onAvailable(network: Network) {
            if (receivedOnAvailable) {
                return
            }

            receivedOnAvailable = true

            mainThreadHandler.post { this@FirstStartWithNoNetworkActivity.showNetworkAcquiredAndFinish() }
        }
    }

    private interface AnimationEndListener : Animator.AnimatorListener {

        override fun onAnimationStart(animation: Animator) {
            // No-op
        }

        override fun onAnimationCancel(animation: Animator) {
            // No-op
        }

        override fun onAnimationRepeat(animation: Animator) {
            // No-op
        }
    }

    companion object {

        private val EXTRA_CONTINUATION_INTENT = "${FirstStartWithNoNetworkActivity::class.java.name}.continuation_intent"
        private const val DELAY_AFTER_ANIMATIONS_MILLIS: Long = 700

        fun createIntentContinuingTo(context: Context, continuationIntent: Intent) =
            Intent(context, FirstStartWithNoNetworkActivity::class.java).apply {
                putExtra(EXTRA_CONTINUATION_INTENT, continuationIntent)
            }
    }
}
